Configure audiences to log warning on invalid events

codemonium opened the pull request:
This change lets applications that use audiences configure the groups that must be present in a user provisioning event for the event to be considered valid. Some applications do not behave correctly if certain groups are missing.

codemonium committed:
Configure audiences to reject invalid events

This change lets applications that use audiences configure the groups
that must be present in a user provisioning event for the event to be
considered valid. Some applications do not behave correctly if certain
groups are missing.

gregblake commented on audiences/lib/audiences/scim/upsert_users_observer.rb:
```diff
@@ -7,6 +7,8 @@ class UpsertUsersObserver < ObserverBase
       subscribe_to "two_percent.scim.replace.Users"
 
       def process
+        validate_user_groups(event_payload.params)
```
Does this logic run when the user's `active` status changes?

I see that "active" is part of the `event_payload.params`. But I'm curious if a change to that attribute triggers a "two_percent.scim.replace.Users" event.

gregblake commented on audiences/lib/audiences/scim/upsert_users_observer.rb:
```diff
@@ -17,6 +19,21 @@ def process
+        expected_types = Audiences.config.required_user_group_types
+
+        return if (expected_types - actual_types).empty?
+
```
Should the UpsertUsersObserver publish another event (e.g. "two_percent.scim.validated.Users") after `validate_user_groups` runs?

gregblake replied:
Done on https://github.com/powerhome/audiences/commit/3289f15b55612d87450f6a1d7a7e70228f21ce7c.

gregblake reviewed (commented):
I think these changes are close to what we need. I added some feedback and questions on comments below.

gregblake committed:
Creates PersistedResourceEvent after persisting the external_user

gregblake committed:
Extracts validations to the model

gregblake committed:
Updates specs: sets active to true on these ExternalUser records

gregblake committed:
Specs now expect ActiveRecord::RecordInvalid on invalid record

gregblake committed:
Inactive users don't require valid groups

gregblake committed:
Previously active users don't require valid groups upon deactivation

gregblake committed:
Adds group association specs

gregblake committed:
Adds test coverage around TwoPercent::ReplaceEvent

gregblake committed:
Adds test coverage

gregblake committed:
Moves UpsertUsersObserver specs into a single file

gregblake committed:
Organizes specs with describe and context blocks

gregblake committed:
Adds helper to reduce duplication

gregblake committed:
Rubocop fixes (part 1)

gregblake committed:
Rubocop pt 2: class size exceeds 100 lines

The addition of the required_group_types method caused the ExternalUser
model to be 101 lines. Rubocop is configured to keep classes at a
maximum of 100 lines.

This change extracts the TERRITORY_ABBBRS constant from the ExternalUser
model. This has two benefits: 1) it resolves the Rubocop violation, and
2) it handles territory abbreviations (which could change on a
per-application basis) as application config. No longer hard coding this
value allows apps that use audiences to do customize the TERRITORY_ABBR
data. For example:

```ruby
Audiences.configure do |config|
  config.territory_abbreviations = { "Custom Territory" => "CT" }
end
```

gregblake committed:
Removes InvalidGroupsError class

The validation logic now uses ActiveRecord::RecordInvalid.

gregblake committed:
Moves model validation specs to external_user_spec.rb

gregblake commented on audiences/lib/audiences/scim/upsert_users_observer.rb:
```diff
@@ -17,6 +19,21 @@ def process
+
+        return if (expected_types - actual_types).empty?
+
+        raise InvalidGroupsError, "User is missing the groups #{expected_types - actual_types}"
```
I don't think it's feasible to resolve this concern on the current PR, because it would likely require changes to the `two_percent` gem.

##### Rationale:

The two_percent gem [fires-and-forgets events](https://github.com/powerhome/power-tools/blob/main/packages/two_percent/app/controllers/two_percent/scim_controller.rb#L5-L15).

The event processing happens asynchronously via the observer. Meanwhile when ActiveRecord::RecordInvalid is raised in the observer, it's logged and re-raised, but the SCIM client already received a 200 OK response.

#### Potential Fix:

Add error handling to [the ScimController actions in the Two Percent gem](https://github.com/powerhome/power-tools/blob/main/packages/two_percent/app/controllers/two_percent/scim_controller.rb). 

Example:

```ruby
module TwoPercent
  class ScimController < ApplicationController
    def create
      TwoPercent::CreateEvent.create(resource: params[:resource_type], params: scim_params)

      head :ok
    rescue ActiveRecord::RecordInvalid => e # it should likely also rescue from other exceptions (e.g. network errors for upstream retry logic)
      render json: { error: e.message }, status: :unprocessable_entity
    end
  end
end
```

gregblake committed:
Adds test coverage for PersistedResourceEvent publishing

gregblake committed:
Fixes a typo in a comment

gregblake committed:
Renames config key to more closely match existing config

config.required_group_types is a better match for config.group_types

gregblake committed:
Retries updating user upon ActiveRecord::RecordInvalid

gregblake committed:
Publishes PersistedResourceEvent after Group is upserted

Some applications need to subscribe_to this event (for example, to add a
new group to a given criteria).

gregblake committed:
Only log a warning if user is missing a required group

gregblake committed:
Simplifies #missing_group_types

Using dependency injection here also makes the tests simpler.

gregblake committed:
Resolves Rubocop issues

gregblake committed:
Merge branch 'main' into JEDI-759

gregblake commented on audiences/lib/audiences/scim/upsert_users_observer.rb:
```diff
@@ -7,12 +7,12 @@ class UpsertUsersObserver < ObserverBase
+
+        return unless valid_group_types?
+
+        Audiences::PersistedResourceEvent.create(resource_type: "Users", params: event_payload.params)
```
This is not a bug. Rails updates the in-memory object without requiring a separate trip to the database:

```ruby
group1 = Audiences::Group.first;
group2 = Audiences::Group.second;
user = Audiences::ExternalUser.new(scim_id: "test-123", user_id: "user-123",
user.groups.count # 0
user.update! updated_attributes
user.groups.count # This is now `2`, without calling `reload`
```

After `update!` is called, the `user.groups` association reflects the new state **without** needing `user.reload`.

gregblake committed:
Adds tests for logging calls

gregblake committed:
Log and raise exceptions in #process

With the changes on this commit, the #process method calls
Audiences.logger.error(e) on any exception, and then raises the
exception. This is consistent with the current behavior on the main
branch, when an actual exception is raised.

Note that the `#valid_group_types?` method still only logs a warning
when there is an user provisioning event where the user is missing one
or more required groups.

Other changes on this commit: #log_upsert_action was extracted from
#process, to keep the Metrics/AbcSize Assignment Branch condition size
under the Rubocop threshold.

gregblake committed:
Updates territory abbreviation config
